src/remote/blobs.go

```diff
@@ -1,241 +1,30 @@
-	Mode   os.FileMode // Only used when receiving blobs, to determine what the output file mode should be
-}
-
 // uploadBlobs uploads a series of blobs to the remote.
 // It handles all the logic around the various upload methods etc.
 // The given function is a callback that receives a channel to send these blobs on; it
 // should close it when finished.
-func (c *Client) uploadBlobs(f func(ch chan<- *blob) error) error {
+func (c *Client) uploadBlobs(f func(ch chan<- *chunker.Chunker) error) error {
 	const buffer = 10 // Buffer it a bit but don't get too far ahead.
-	chIn := make(chan *blob, buffer)
-	chOut := make(chan *blob, buffer)
+	ch := make(chan *chunker.Chunker, buffer)
 	var g errgroup.Group
-	g.Go(func() error { return f(chIn) })
-	g.Go(func() error { return c.reallyUploadBlobs(chOut) })
-	// We use this to deduplicate outgoing blobs; there are cases of repos with hundreds of
-	// copies of the same file, and there's no point checking over and over.
-	// We don't do it at a more global level but this is *probably* sufficient for now.
-	seen := map[string]struct{}{}
-
-	// This function filters a set of blobs through FindMissingBlobs to find out which
-	// ones we actually need to upload. The assumption is that most of the time the
-	// server will already have a subset of the blobs we're going to upload so we don't
-	// want to send them all again.
-	filter := func(blobs []*blob) {
-		req := &pb.FindMissingBlobsRequest{
-			InstanceName: c.instance,
-			BlobDigests:  make([]*pb.Digest, 0, len(blobs)),
-		}
-		m := make(map[string]*blob, len(blobs))
-		for _, b := range blobs {
-			if _, present := seen[b.Digest.Hash]; !present {
-				seen[b.Digest.Hash] = struct{}{}
-				req.BlobDigests = append(req.BlobDigests, b.Digest)
-				m[b.Digest.Hash] = b
-			}
-		}
-		ctx, cancel := context.WithTimeout(context.Background(), c.reqTimeout)
-		defer cancel()
-		resp, err := c.client.FindMissingBlobs(ctx, req)
-		if err != nil {
-			log.Warning("Error filtering blobs for remote execution: %s", err)
-			// Continue and send all of these, it is not necessarily fatal (although it
-			// will probably blow up later, but easier not to handle that error here)
-			for _, b := range blobs {
-				chOut <- b
-			}
-			return
-		}
-		for _, d := range resp.MissingBlobDigests {
-			chOut <- m[d.Hash]
-		}
-	}
-
-	// Buffer them up a bit, the request supports checking multiple at once which is
-	// likely more efficient than doing them all one at a time.
-	blobs := make([]*blob, 0, buffer)
-	for b := range chIn {
-		// The actual hash might or might not be set.
-		if len(b.Digest.Hash) == 0 {
-			h, err := c.state.PathHasher.Hash(b.File, false, true)
-			if err != nil {
-				return err
-			}
-			b.Digest.Hash = hex.EncodeToString(h)
-		}
-		blobs = append(blobs, b)
-		if len(blobs) == buffer {
-			filter(blobs)
-			blobs = blobs[:0]
-		}
-	}
-	if len(blobs) != 0 {
-		filter(blobs)
+	g.Go(func() error { return f(ch) })
+	chomks := []*chunker.Chunker{}
```

Comment:
Perhaps better english would be "chunks"?

Reply:
The TM memes have infected me to the point where this makes perfect sense ;-)